src/bloom/hierarchical.rs

```diff
@@ -0,0 +1,335 @@
+use std::iter::FromIterator;
+
+use super::*;
+
+/// Container for types which is support bloom filtering
+#[derive(Debug)]
+pub struct HierarchicalBloom<Child>
+where
+    Child: BloomProvider + Send + Sync,
+{
+    inner: HierarchicalBloomInner,
+    children: Vec<Child>,
+    group_size: usize,
+}
+
+#[derive(Debug)]
+enum HierarchicalBloomInner {
+    /// Inner filters
+    Vec(Option<Bloom>, Vec<HierarchicalBloomInner>),
+    /// Leaf
+    Item(usize),
+}
+
+impl Default for HierarchicalBloomInner {
+    fn default() -> Self {
+        Self::Vec(Default::default(), Default::default())
+    }
+}
+
+impl<Child> Default for HierarchicalBloom<Child>
+where
+    Child: BloomProvider + Send + Sync,
+{
+    fn default() -> Self {
+        Self {
+            inner: Default::default(),
+            children: Default::default(),
+            group_size: 8,
+        }
+    }
+}
+
+impl HierarchicalBloomInner {
+    #[async_recursion::async_recursion]
+    async fn check_filter<Child, Key>(&self, children: &[Child], item: &Key) -> Result<Option<bool>>
+    where
+        Child: BloomProvider<Key = Key> + Send + Sync,
+        Key: Send + Sync + AsRef<[u8]> + ?Sized,
+    {
+        match self {
```

Comment:
The logic seems to be wrong. Bloom filter can detect that key is definitely not presented (`false`). If it returns `true`, that means that key is presented with some probability. Thus, you must immediately return `false` when current level filter also returns `false`. If the current level filter returns `true`, then you need to go down the hierarchy